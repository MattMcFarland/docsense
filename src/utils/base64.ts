export const encode = (str: string) => {
  const buffer = Buffer.from(str, 'utf8');
  return buffer.toString('base64');
};
export const decode = (str: string) => {
  const buffer = Buffer.from(str, 'base64');
  return buffer.toString('utf8');
};
