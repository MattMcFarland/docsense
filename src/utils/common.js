// @flow

/**
 * Send a status message while in a promise chain
 * @param {string} msg
 * @returns {Passthrough} - context of the promise chain is unchanged.
 */
export const status = (msg: string) => (context: Context): any => {
  log.info('status', msg)
  return context
}

export const logMap = (prefix: string) => (items: string[]): string[] => {
  items.forEach(item => log.info(prefix, item))
  return items
}

/**
 * Changes context in promise chain
 * @param {*} newContext next thing in promise chain will focus on this
 */
export const setContext = (newContext: Context) => (): Promise<any> =>
  Promise.resolve(newContext)

/**
 * Wipes the context to undefined, useful for controlling flow
 * @returns {void}
 */
export const voidContext = setContext(undefined)

/**
 * Writes the context of a promise chain to the given key as an entry ( [key, context] )
 * @param {string} key
 * @returns {entry}
 */
export const contextAsEntry = <K, V>(key: K) => (context: V): Entry<K, V> => [
  key,
  context,
]

/**
 * Converts entries to a POJO
 * @param {entry[]} entries
 * @returns {POJO} converted entries
 */
export const convertEntriesToObject = <K, V>(entries: Entry<K, V>[]) =>
  entries.reduce(
    (obj, [key, value]) => Object.defineProperty(obj, key, { value }),
    {}
  )

/**
 * Extracts values from entries
 * @param {entry[]} entries
 * @returns {V[]} values
 */
export const extractValuesFromEntries = <K, V>(entries: Entry<K, V>[]): V[] =>
  entries.map(([, value]: Entry<K, V>) => (value: V))

/**
 * Logs error, exits 1
 * @param {Error} err error object
 * @returns {void}
 */
export const fatalError = (err: Error): void => {
  log.error('ERR', err)
  process.exit(1)
}
/**
 * Logs the current context to the console
 * @param {*} context log this
 * @returns {*} context
 */
export const logContext = (context: Context): any => {
  log.info('ctx', context)
  return context
}

/**
 * Removes duplicates from the given array
 * @param {Array} arr dedupe this array
 * @returns {Array} deduped
 */
export const dedupe = (arr: Array<any>): Array<any> =>
  arr.reduce((x, y) => (x.includes(y) ? x : [...x, y]), [])

/**
 * Flattens an array that is one level deep.
 * @param {Array} arr flatten this array
 * @returns {Array} deduped
 */
export const flatten = (arr: Array<any>): Array<any> => [].concat(...arr)
