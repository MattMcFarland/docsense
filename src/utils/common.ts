const log = global.log;

type Entry = [string, any];
/**
 * Send a status message while in a promise chain
 * @param {string} msg
 * @returns {Passthrough} - context of the promise chain is unchanged.
 */
export const status = (msg: string) => (context: any): any => {
  log.info('status', msg);
  return context;
};

export const logMap = (prefix: string) => (items: string[]): string[] => {
  items.forEach(item => log.info(prefix, item));
  return items;
};

/**
 * Changes context in promise chain
 * @param {*} newContext next thing in promise chain will focus on this
 */
export const setContext = (newContext: any) => (): Promise<any> =>
  Promise.resolve(newContext);

/**
 * Wipes the context to undefined, useful for controlling flow
 * @returns {void}
 */
export const voidContext = setContext(undefined);

/**
 * Writes the context of a promise chain to the given key as an entry ( [key, context] )
 * @param {string} key
 * @returns {entry}
 */
export const contextAsEntry = (key: string) => (context: any): Entry => [
  key,
  context,
];

/**
 * Converts entries to a POJO
 * @param {entry[]} entries
 * @returns {POJO} converted entries
 */
export const convertEntriesToObject = (entries: Entry[]) =>
  entries.reduce(
    (obj, [key, value]) => Object.defineProperty(obj, key, { value }),
    {}
  );

/**
 * Logs error, exits 1
 * @param {Error} err error object
 * @returns {void}
 */
export const fatalError = (err: Error): void => {
  log.error('ERR', err);
  process.exit(1);
};
/**
 * Logs the current context to the console
 * @param {*} context log this
 * @returns {*} context
 */
export const logContext = (context: any): any => {
  log.info('ctx', context);
  return context;
};

/**
 * Removes duplicates from the given array
 * @param {Array} arr dedupe this array
 * @returns {Array} deduped
 */
export const dedupe = (arr: any[]): any[] =>
  arr.reduce((x, y) => (x.includes(y) ? x : [...x, y]), []);

/**
 * Flattens an array that is one level deep.
 * @param {Array} arr flatten this array
 * @returns {Array} deduped
 */
export const flatten = (arr: any[]): any[] => [].concat(...arr);
