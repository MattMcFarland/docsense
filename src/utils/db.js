// @flow

// /**
//  * Returns the value from database
//  * @param {K} key
//  * @returns {V} value
//  */
// export const fromDB = (key: K): V =>
//   cache
//     .get("db")
//     .get(key)
//     .value();

// /**
//  * Writes key value to database
//  * @param {K} key write this key
//  * @param {V} value with this value
//  * @returns {void}
//  */
// export const toDB = (key: K, value: V): Promise<void> =>
//   Promise.resolve(
//     cache
//       .get("db")
//       .get(key)
//       .assign(value)
//       .write()
//   );
