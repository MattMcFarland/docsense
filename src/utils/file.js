// @flow

import { resolve as resolvePath } from 'path'
import { promisify } from 'util'
import fs from 'fs'
import glob from 'glob'

const readFile = promisify(fs.readFile)

/**
 * Processes a glob pattern to an array of files
 * @param {string} pattern pattern to resolve
 * @returns {Promise<string>} all matches found from the pattern
 * @throws {Error} If nothing is found
 * @see processAllGlobPatterns
 */
export const processGlobPattern = (pattern: string): Promise<string> =>
  new Promise((resolve, reject) => {
    glob(pattern, (err, matches) => {
      if (err) return reject(err)
      return resolve(matches)
    })
  })

/**
 * Processes an array of glob patterns to an array of files
 * @param {string} pattern pattern to resolve
 * @returns {Promise<string[]>} all matches found from the pattern
 * @throws {Error} If nothing is found
 * @uses processGlobPattern
 */
export const processAllGlobPatterns = (patterns: string[]): Promise<string[]> =>
  Promise.all(patterns.map(processGlobPattern))

/**
 * Converts resolves the full path of the filepath relative to the current working directory.
 * @param {string} relativePath filepath to resolve
 * @returns {string} fullpath
 * @uses resolvePath
 */
export const resolvePathFromCWD = (relativePath: string): string =>
  resolvePath(process.cwd(), relativePath)

/**
 * Resolves the full path all file patterns relative to the current working directory.
 * @param {string[]} relativePaths array of relative filepaths to resolve
 * @returns {string[]} fullpaths
 * @uses resolvePathFromCWD
 */
export const resolveAllFilePathsFromCWD = (relativePaths: string[]): string[] =>
  relativePaths.map(resolvePathFromCWD)

/**
 * Opens file, reads it, then creates an entry where the
 * filepath is key, and the raw data is the value.
 * @param {string} filepath path to be read by fs.readFile
 * @returns {Promise<entry>} [filepath, ast]
 * @uses openFileForReading
 * @uses contextAsEntry
 */
export const safelyReadFile = (filepath: string): Promise<entry> =>
  resolveFile(filepath).then(openFileForReading)

/**
 * Only attempts to read real files, discarding directories, etc.
 * @param {string} filepath path to be read by fs.readFile
 * @returns {Promise<string>} The resolved file path
 * @see readFiles
 * @see parseFiles
 */
export const resolveFile = (filepath: string): Promise<string> => {
  return new Promise((resolve, reject) => {
    fs.stat(filepath, (err, stats) => {
      if (err) return reject(err)
      if (stats.isFile()) {
        console.log('resolve', filepath)
        return resolve(filepath)
      }
      console.warn('skip', filepath)
      return
    })
  })
}

/**
 * performs fs.readfile at utf-8 encoding on the given filepath
 * @param {string} filepath - fs.readfile this
 * @returns {string} utf-8 content
 */
export const openFileForReading = (filepath: string) =>
  readFile(filepath, 'utf-8')

/**
 * Read the contents of all files in the given array
 * @param {string[]} filesArray - array of full paths to files
 * @returns {Promise<entry[]>} - A promise when all files have been read
 * @uses safelyReadFile
 */
export const readFiles = (filesArray: string[]): Promise<entry[]> =>
  Promise.all(filesArray.map(safelyReadFile))
