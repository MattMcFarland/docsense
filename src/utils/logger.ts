import logger from 'npmlog';
import pkg from '../package.json';
import assert from 'assert';

let initialized = false;

export const init = () => {
  assert(!initialized, 'logger already initialized');
  logger.heading = 'docsense';
  logger.info('using', `docsense@${pkg.version}`);
  logger.addLevel('success', 2000, { fg: 'green', bold: true });
  initialized = true;
};

export const log = logger;
