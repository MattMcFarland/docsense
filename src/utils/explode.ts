import { DirectoryModel, FileIdRecord } from '../core-plugins/directory';
import { decode, encode } from './base64';

export interface DirectoryTree {
  [key: string]: DirectoryTree[] | string[];
}
/**
 * Explodes the dir_collection into a tree.
 */
export default function explode(
  dir_collection: DirectoryModel[]
): DirectoryTree {
  const ds = dir_collection.map(d => d.directory_id + '/');
  const ps = dir_collection.reduce((acc: string[], dirModel): string[] => {
    if (dirModel.files) {
      dirModel.files.forEach((f: FileIdRecord) => {
        const filepath = decode(f.file_id);
        acc.push(filepath);
      });
    }
    return acc;
  }, []);

  const paths = [...ds, ...ps];

  /**
   * Extract a filename from a path
   * @param dir_collection dir_collection from database
   * @author Stephane Janicaud
   * @link https://stackoverflow.com/questions/43431829/split-array-of-file-paths-into-hierarchical-object-in-javascript/43432913#43432913
   */
  function getFilename(path: string): string {
    return path
      .split('/')
      .filter(value => {
        return value && value.length;
      })
      .reverse()[0];
  }

  /**
   * Find sub paths
   * @param path
   * @author Stephane Janicaud
   * @link https://stackoverflow.com/questions/43431829/split-array-of-file-paths-into-hierarchical-object-in-javascript/43432913#43432913
   */
  function findSubPaths(path: string): string[] {
    // slashes need to be escaped when part of a regexp
    const rePath = path.replace('/', '\\/');
    const re = new RegExp('^' + rePath + '[^\\/]*\\/?$');
    return paths.filter(i => {
      return i !== path && re.test(i);
    });
  }

  /**
   * Build tree recursively
   * @param path
   * @author Stephane Janicaud
   * @link https://stackoverflow.com/questions/43431829/split-array-of-file-paths-into-hierarchical-object-in-javascript/43432913#43432913
   */
  function buildTree(path: string = ''): DirectoryTree {
    const nodeList: any = [];
    findSubPaths(path).forEach(subPath => {
      const nodeName = getFilename(subPath);
      if (/\/$/.test(subPath)) {
        const node: any = {};
        node[nodeName] = buildTree(subPath);
        nodeList.push(node);
      } else {
        nodeList.push(encode(subPath));
      }
    });
    return nodeList;
  }

  // Build tree from root
  return buildTree();
}
