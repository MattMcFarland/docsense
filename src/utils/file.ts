import * as FS from 'fs';
import * as Glob from 'glob';
import * as MkDirP from 'mkdirp';
import * as Path from 'path';
import { promisify } from 'util';

import { log } from '../utils/logger';

const readFile = promisify(FS.readFile);

/**
 * Processes a glob pattern to an array of files
 * @param {string} pattern pattern to resolve
 * @returns {Promise<string>} all matches found from the pattern
 * @see processAllGlobPatterns
 */
export const processGlobPattern = (pattern: string): Promise<string[]> =>
  new Promise((resolve, reject) =>
    Glob(pattern, (err, matches) => (err ? reject(err) : resolve(matches)))
  );

/**
 * Processes an array of glob patterns to an array of files
 * @param {string[]} patterns patterns to resolve
 * @returns {Promise<string[]>} all matches found from the pattern
 * @uses processGlobPattern
 */
export const processAllGlobPatterns = (
  patterns: string[]
): Promise<string[][]> => Promise.all(patterns.map(processGlobPattern));

/**
 * Converts resolves the full path of the filepath relative to the current working directory.
 * @param {string} relativePath filepath to resolve
 * @returns {string} fullpath
 */
export const resolvePathFromCWD = (relativePath: string): string =>
  Path.posix.resolve(process.cwd(), relativePath);

/**
 * Resolves the full path all file patterns relative to the current working directory.
 * @param {string[]} relativePaths array of relative filepaths to resolve
 * @returns {string[]} fullpaths
 * @uses resolvePathFromCWD
 */
export const resolveAllFilePathsFromCWD = (relativePaths: string[]): string[] =>
  relativePaths.map(resolvePathFromCWD);

/**
 * Opens file, reads it, then creates an entry where the
 * filepath is key, and the raw data is the value.
 * @param {string} filepath path to be read by fs.readFile
 * @returns {Promise<entry>} [filepath, ast]
 * @uses openFileForReading
 * @uses contextAsEntry
 */
export const safelyReadFile = (filepath: string): Promise<string> =>
  resolveFile(filepath).then(openFileForReading);

/**
 * Only attempts to read real files, discarding directories, etc.
 * @param {string} filepath path to be read by fs.readFile
 * @returns {Promise<string>} The resolved file path
 * @see readFiles
 * @see parseFiles
 */
export const resolveFile = (filepath: string): Promise<string> => {
  return new Promise((resolve, reject) => {
    FS.stat(filepath, (err, stats) => {
      if (err) {
        return reject(err);
      }
      if (stats.isFile()) {
        log.info('read', filepath);
        return resolve(filepath);
      }
      log.warn('skip', filepath);
    });
  });
};

/**
 * performs fs.readfile at utf-8 encoding on the given filepath
 * @param {string} filepath - fs.readfile this
 * @returns {string} utf-8 content
 */
export const openFileForReading = (filepath: string): Promise<string> =>
  readFile(filepath, 'utf-8');

/**
 * Read the contents of all files in the given array
 * @param {string[]} filesArray - array of full paths to files
 * @returns {Promise<entry[]>} - A promise when all files have been read
 * @uses safelyReadFile
 */
export const readFiles = (filesArray: string[]): Promise<string[]> =>
  Promise.all(filesArray.map(safelyReadFile));

/**
 * Given an array of directory items (typically from fs.readdir), reduce it
 * down to just js files
 * @param {string[]} directory
 */
export const reduceDirectoryToJSFiles = (directory: string[]) =>
  directory.reduce((validFiles: string[], name: string) => {
    if (name.endsWith('.js')) {
      validFiles.push(name);
    }
    return validFiles;
  }, []);

/**
 * scan a directory using fs.readdir
 * @param {string} directoryPath
 */
export const scanDirectory = (directoryPath: string) => (): Promise<string[]> =>
  new Promise((resolve, reject) => {
    FS.readdir(directoryPath, (err, directory) => {
      if (err) {
        return reject(err);
      }
      return resolve(directory);
    });
  });

/**
 * Given the current promise chain context, resolve the paths from the relativePath
 * arguments
 * @param {arguments} relativePath
 */
export const resolveContextRelativePaths = (...relativePath: string[]) => (
  filenames: string[]
): string[] =>
  filenames.map((filename: string): string =>
    Path.posix.join(...relativePath, filename)
  );

/**
 * Evals the given function over each of the files within the given directoryPath
 * @param directoryPath path to iterate over
 * @param fn function called on each file within the directoryPath
 */
export const withAllFiles = (
  directoryPath: string,
  fn: (data: string, filename: string, index: number) => void
) => {
  const files = FS.readdirSync(directoryPath);
  files.forEach((filename, index) => {
    const data = FS.readFileSync(Path.resolve(directoryPath, filename), 'utf8');
    fn(data, filename, index);
  });
};

/**
 * Creates file from the given path, even if the path does not exist
 * Similar to running `mkdir -p [path] ; touch [path]
 * @param target target path and file
 * @param data written to the new file
 */
export const createFile = (target: string, data: string | Buffer) =>
  new Promise((resolve, reject) => {
    const dir = Path.dirname(target);

    MkDirP(dir, mkDirError => {
      if (mkDirError) return reject(mkDirError);
      FS.writeFile(target, data, writeError => {
        if (writeError) return reject(writeError);
        if (!writeError) return resolve(true);
      });
    });
  });
