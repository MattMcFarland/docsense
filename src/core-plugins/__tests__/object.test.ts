import { testFactory } from './utils';

const objectPlugin = require('../object');
const suites = {
  '>': [
    'const foo = {}',
    `const obj = {
      foo: bar,
      baz: bez,
      fn() { },
      thing: () => {},
      num: 100,
      numstr: "onehundred",
      dfn: function () {}
    }`,
    `
    /**
     * Super hero Record.
     * @property {string} name - super hero name
     * @property {string} power - the hero ability
     * @property {method} attack - attack
     * @property {number} level - hero level
     */
    const superHero = {
      name: 'Batman',
      power: 'Super Rich',
      attack() {},
      level: 100,
    }`,
    'const foo = { bar: { name: "baz" }}',
  ],
};

describe('Core: object', () =>
  testFactory({
    plugin: objectPlugin,
    suites,
  }));
