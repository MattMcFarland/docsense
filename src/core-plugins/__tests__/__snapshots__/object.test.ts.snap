// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core: object > 
    /**
     * Super hero Record.
     * @property {string} name - super hero name
     * @property {string} power - the hero ability
     * @property {method} attack - attack
     * @property {number} level - hero level
     */
    const superHero = {
      name: 'Batman',
      power: 'Super Rich',
      attack() {},
      level: 100,
    } 1`] = `
Array [
  Object {
    "file_id": "__TEST__",
    "jsdoc": Array [
      Object {
        "description": "Super hero Record.",
        "tags": Array [
          Object {
            "description": "super hero name",
            "name": "name",
            "title": "property",
            "type": Object {
              "name": "string",
              "type": "NameExpression",
            },
          },
          Object {
            "description": "the hero ability",
            "name": "power",
            "title": "property",
            "type": Object {
              "name": "string",
              "type": "NameExpression",
            },
          },
          Object {
            "description": "attack",
            "name": "attack",
            "title": "property",
            "type": Object {
              "name": "method",
              "type": "NameExpression",
            },
          },
          Object {
            "description": "hero level",
            "name": "level",
            "title": "property",
            "type": Object {
              "name": "number",
              "type": "NameExpression",
            },
          },
        ],
      },
    ],
    "object_id": "superHero",
  },
]
`;

exports[`Core: object > const foo = { bar: { name: "baz" }} 1`] = `
Array [
  Object {
    "file_id": "__TEST__",
    "jsdoc": undefined,
    "object_id": "foo",
  },
  Object {
    "file_id": "__TEST__",
    "jsdoc": undefined,
    "object_id": "bar",
  },
]
`;

exports[`Core: object > const foo = {} 1`] = `
Array [
  Object {
    "file_id": "__TEST__",
    "jsdoc": undefined,
    "object_id": "foo",
  },
]
`;

exports[`Core: object > const obj = {
      foo: bar,
      baz: bez,
      fn() { },
      thing: () => {},
      num: 100,
      numstr: "onehundred",
      dfn: function () {}
    } 1`] = `
Array [
  Object {
    "file_id": "__TEST__",
    "jsdoc": undefined,
    "object_id": "obj",
  },
]
`;
